contrib/parseq-retry/src/main/java/com/linkedin/parseq/retry/RetriableTask.java

```diff
@@ -0,0 +1,184 @@
+          ErrorClassification errorClassification = _policy.getErrorClassifier().apply(task.getError());
+          retry(attempt + 1, Failure.of(task.getError()), errorClassification, recoveryContext, recoveryResult);
+        } else {
+          // Successful task might still need to be retried if its result is not acceptable according to the policy.
+          T taskResult = task.get();
+          ResultClassification resultClassification = _policy.getResultClassifier().apply(taskResult);
+          if (resultClassification == ResultClassification.UNACCEPTABLE) {
+            // Unacceptable result -> should retry.
+            retry(attempt + 1, Success.of(taskResult), resultClassification.getStatus(), recoveryContext, recoveryResult);
+          } else {
+            // Acceptable result -> we are done.
+            recoveryResult.done(taskResult);
+          }
+        }
+
+        return recoveryResult;
+      });
+
+      // Recovery task should run immediately after the original task to process its outcome.
+      recovery.setPriority(Priority.MAX_PRIORITY);
+      recovery.getShallowTraceBuilder().setSystemHidden(true);
+      Promises.propagateResult(recovery, result);
+      context.after(task).run(recovery);
+      context.run(task);
+
+      return result;
+    });
+  }
+
+  /** Invoke event monitors and schedule a retry if policy allows. */
+  private void retry(int attempt, Try<T> outcome, ErrorClassification errorClassification, Context recoveryContext, SettablePromise<T> recoveryResult) {
+    long backoffTime = _policy.getBackoffPolicy().nextBackoff(attempt, outcome);
+
+    if (errorClassification.isFatal()) {
+      // For fatal errors there are no retries.
+      _policy.getEventMonitor().interrupted(_name, outcome, attempt);
+      recoveryResult.fail(outcome.isFailed() ? outcome.getError() : new RetryFailureException("Retry aborted because of unacceptable task result: " + outcome.get()));
+    } else if (_policy.getTerminationPolicy().shouldTerminate(attempt, System.currentTimeMillis() - _startedAt + backoffTime)) {
+      // Retry policy commands that no more retries should be done.
+      _policy.getEventMonitor().aborted(_name, outcome, attempt);
+      recoveryResult.fail(outcome.isFailed() ? outcome.getError() : new RetryFailureException("Retry aborted because of unacceptable task result: " + outcome.get()));
+    } else {
+      // Schedule a new retry task after a computed backoff timeout.
+      _policy.getEventMonitor().retrying(_name, outcome, attempt, backoffTime, errorClassification.isSilent());
+      Task<T> retryTask = wrap(attempt);
+      Promises.propagateResult(retryTask, recoveryResult);
+      recoveryContext.createTimer(backoffTime, TimeUnit.MILLISECONDS, retryTask);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected Promise<? extends T> run(Context context) {
+    _startedAt = System.currentTimeMillis();
+    Task<T> task = wrap(0);
```

Comment:
I do have explicit documentation about this, please check javadocs on the publicly available functions. Did I miss it somewhere?